core/amber/src/main/scala/edu/uci/ics/texera/web/resource/WorkflowResultService.scala

```diff
@@ -0,0 +1,217 @@
+package edu.uci.ics.texera.web.resource
+
+import com.fasterxml.jackson.annotation.{JsonTypeInfo, JsonTypeName}
+import com.fasterxml.jackson.databind.node.ObjectNode
+import edu.uci.ics.amber.engine.architecture.controller.ControllerEvent.WorkflowResultUpdate
+import edu.uci.ics.amber.engine.architecture.principal.OperatorResult
+import edu.uci.ics.amber.engine.common.tuple.ITuple
+import edu.uci.ics.texera.web.model.event.TexeraWebSocketEvent
+import edu.uci.ics.texera.web.resource.WorkflowResultService.{
+  PaginationMode,
+  SetDeltaMode,
+  SetSnapshotMode,
+  WebOutputMode,
+  WebPaginationUpdate,
+  WebResultUpdate,
+  defaultPageSize,
+  webDataFromTuple
+}
+import edu.uci.ics.texera.web.resource.WorkflowResultService.calculateDirtyPageIndices
+import edu.uci.ics.texera.web.resource.WorkflowWebsocketResource.send
+import edu.uci.ics.texera.workflow.common.workflow.WorkflowCompiler
+import edu.uci.ics.texera.workflow.operators.sink.SimpleSinkOpDesc
+import edu.uci.ics.texera.workflow.common.IncrementalOutputMode.{SET_DELTA, SET_SNAPSHOT}
+import edu.uci.ics.texera.workflow.common.tuple.Tuple
+
+import javax.websocket.Session
+import scala.collection.mutable
+
+object WorkflowResultService {
+
+  val defaultPageSize: Int = 10
+
+  /**
+    * Behavior for different web output modes:
+    *  - PaginationMode   (used by view result operator)
+    *     - send new number of tuples and dirty page index
+    *  - SetSnapshotMode  (used by visualization in snapshot mode)
+    *     - send entire snapshot result to frontend
+    *  - SetDeltaMode     (used by visualization in delta mode)
+    *     - send incremental delta result to frontend
+    */
+  @JsonTypeInfo(use = JsonTypeInfo.Id.NAME, include = JsonTypeInfo.As.PROPERTY, property = "type")
+  sealed abstract class WebOutputMode extends Product with Serializable
+  @JsonTypeName("PaginationMode")
+  final case class PaginationMode() extends WebOutputMode
+  @JsonTypeName("SetSnapshotMode")
+  final case class SetSnapshotMode() extends WebOutputMode
+  @JsonTypeName("SetDeltaMode")
+  final case class SetDeltaMode() extends WebOutputMode
+
+  /**
+    * The result update of one operator that will be sent to the frontend.
+    * Can be either WebPaginationUpdate (for PaginationMode)
+    *            or WebDataUpdate (for SetSnapshotMode or SetDeltaMode)
+    */
+  sealed abstract class WebResultUpdate extends Product with Serializable
+
+  case class WebPaginationUpdate(
+      mode: PaginationMode,
+      totalNumTuples: Int,
+      dirtyPageIndices: List[Int]
+  ) extends WebResultUpdate
+
+  case class WebDataUpdate(mode: WebOutputMode, table: List[ObjectNode], chartType: Option[String])
```

Comment:
we have too many names, `WebData`, `WebResult`, `WebOutput`... I wonder if we actually need all those variants, or maybe we can combine some of them. It is not intuitive to me the relationship between the above three names: are they parallel or are they have an ISA relationship (just from the name, without looking at code)? 

Also I personally do not like `WebData` this particular name. maybe it is because data is a very general term. in the engine we can use data message vs control message, but in the frontend and web server, when there is no control, data becomes very broad term.

Reply:
Yes we do need all these variants. 
`WebResultUpdate` is the base class for the ADT and `WebPaginationUpdate`/`WebDataUpdate` are the only two types. 
I also don't like `WebDataUpdate`, can you help suggest a better name?